lib/helpers/signed-aws-es-fetch.js

```diff
@@ -0,0 +1,124 @@
+const aws4 = require('aws4');
+const nodeFetch = require('node-fetch');
+const logger = require('@financial-times/n-logger').default;
+const resolveCname = require('util').promisify(require('dns').resolveCname);
+const memoize = require('lodash/memoize');
+
+const memoizedResolveCname = memoize(resolveCname);
+
+const defaultAwsKeys = {
+	awsKeys: ['ES_AWS_ACCESS_KEY', 'AWS_ACCESS_KEY', 'AWS_ACCESS_KEY_ID'],
+	awsSecretKeys: ['ES_AWS_SECRET_ACCESS_KEY', 'AWS_SECRET_ACCESS_KEY'],
+	tokenKeys: ['ES_AWS_SESSION_TOKEN', 'AWS_SESSION_TOKEN']
+};
+
+module.exports = async function (url, opts, creds) {
+	opts = opts || {};
+	url = await resolveUrlAndHost(url);
+	return signedFetch(url, opts, creds);
+};
+
+function getURLObject(url) {
+	try {
+		const urlObject = new URL(url);
+		return urlObject;
+	} catch (error) {
+		logger.error({
+			event: 'Invalid URL',
+			data: {
+				url
+			}
+		});
+		throw error;
+	}
+}
+
+async function resolveValidHost(urlObject) {
+	try {
+		const hosts = await memoizedResolveCname(urlObject.host);
+		if (hosts && hosts[0]) {
+			return hosts[0];
+		}
+	} catch (error) {}
+	logger.error({
+		event: 'Invalid Host',
+		data: {
+			url: urlObject
+		}
+	});
+	throw new Error('Invalid Host');
+}
+
+function isActiveDnsResolution() {
+	return (
+		!process.env.AWS_SIGNED_FETCH_DISABLE_DNS_RESOLUTION ||
+		process.env.AWS_SIGNED_FETCH_DISABLE_DNS_RESOLUTION === 'false'
+	);
+}
+
+async function resolveUrlAndHost(url) {
+	const urlObject = getURLObject(url);
+	if (
+		!/\.es\.amazonaws\.com$/.test(urlObject.host) &&
+		isActiveDnsResolution()
+	) {
+		const host = await resolveValidHost(urlObject);
+		url = url.replace(urlObject.host, host);
+	}
+	return url;
+}
+
+function defaultAwsCredentials(keys) {
+	const keyName = defaultAwsKeys[keys].find((keyName) => process.env[keyName]);
```

Comment:
One thing that I find interesting about this function is that we could end up with a mix of keys and secrets - for example `ES_AWS_ACCESS_KEY` used with `AWS_SECRET_ACCESS_KEY` which wouldn't work.

I see this is in the original code though and it hasn't broken before so let's not change it right now. Just found it funny.

Reply:
Yep, here there are the official key names according to the official @aws-sdk/credential-provider. 
Maybe some repositories have several AWS keys and they need different names for each of them.
I'll check this one and fix it accordingly.
```
import { CredentialProvider } from "@aws-sdk/types";
export declare const ENV_KEY = "AWS_ACCESS_KEY_ID";
export declare const ENV_SECRET = "AWS_SECRET_ACCESS_KEY";
export declare const ENV_SESSION = "AWS_SESSION_TOKEN";
export declare const ENV_EXPIRATION = "AWS_CREDENTIAL_EXPIRATION";
/**
 * Source AWS credentials from known environment variables. If either the
 * `AWS_ACCESS_KEY_ID` or `AWS_SECRET_ACCESS_KEY` environment variable is not
 * set in this process, the provider will return a rejected promise.
 */
export declare const fromEnv: () => CredentialProvider;
```